internal/pkg/agent/application/coordinator/coordinator.go

```diff
@@ -445,12 +448,23 @@ func (c *Coordinator) Upgrade(ctx context.Context, version string, sourceURI str
 
 	// override the overall state to upgrading until the re-execution is complete
 	c.SetOverrideState(agentclient.Upgrading, fmt.Sprintf("Upgrading to version %s", version))
-	cb, err := c.upgradeMgr.Upgrade(ctx, version, sourceURI, action, skipVerifyOverride, skipDefaultPgp, pgpBytes...)
+
+	// initialize upgrade details
+	actionID := ""
+	if action != nil {
+		actionID = action.ActionID
+	}
+	det := details.NewDetails(version, details.StateRequested, actionID)
+	det.RegisterObserver(c.setUpgradeDetails)
```

Comment:
This is the real problem behind the race you hit with `stateNeedsRefresh` -- the comments for `setUpgradeDetails` say `// Must be called on the main Coordinator goroutine.` but this observer can be called from elsewhere, which means not just the refresh flag but the upgrade details themselves are being accessed unprotected during an update.

To do this safely, imitate the pattern in `SetOverrideState` / `setOverrideState`, where the former is a public method that forwards to the appropriate update channel in `Coordinator`, and the latter is the internal method called directly from the main run loop.